### PR TITLE
2154 refactoring bvh cpp classes

### LINKED ISSUE #2154: Refactoring BVH C++ classes
- [x] memory leak
- [x] code style
- [x] optimize tree building
- [x] optimize ray intersection algo (partial work)
- [x] refactoring build and ray intersection api
- [x] implement api with another backend : https://github.com/madmann91/bvh
- [x] support range element and topological faces
- [x] option for bvh build and ray intersection  (i.e. quality, robust)

### PR DESCRIPTION
**Pull request by vincentchabannes**
- [ ] Have you checked to ensure there aren't other open [Pull Requests](https://github.com/feelpp/feelpp/pulls) for the same change?
- [ ] Have you added an explanation of what your changes do and why you'd like us to include them?
- [ ] Have you written new tests for your changes?
- [ ] Have you successfully run the Feel++ testsuite with your changes locally?
- [ ] Have you written Doxygen comments in your contribution ?

-----
- closes #2154

### CODE REVIEW

**Comment by vincentchabannes** [comment]
@prudhomm it's ready for review. Huge perf with new BVH (thanks to the third party bvh lib). Now the test `feelpp_test_viewfactor_raytracing` is run 4s instead of 220s (the shading_mask app is also very fast now).

File: feelpp/feel/feelmesh/bvh.hpp [hunk 1]
```diff
@@ -1,303 +1,438 @@
...
+    std::vector<BVHPrimitiveInfo> M_primitiveInfo;
+    BVHEnum::Quality M_quality = BVHEnum::Quality::High;
+};
 
-        class BVHNode
+
+
+//! @brief implementation of BVH tool with an external third party
+template <typename MeshEntityType>
+class BVH_ThirdParty : public BVH<MeshEntityType>
```
Comment (prudhomm) [inline feelpp/feel/feelmesh/bvh.hpp:176]: We don't have _ in class names, how about BVHThirdParty ? or BVHExternal  ?

File: feelpp/feel/feelmesh/bvh.hpp [hunk 2]
```diff
@@ -1,303 +1,438 @@
...
+        }
+private:
+    std::unique_ptr<backend_bvh_type> M_bvh;
+    std::vector<backend_precompute_triangle_type> M_precomputeTriangle;
+};
+
+
+//! @brief in house implementation of BVH tool
+template <typename MeshEntityType>
+class BVH_InHouse : public BVH<MeshEntityType>
```
Comment (prudhomm) [inline feelpp/feel/feelmesh/bvh.hpp:314]: How about BVHInHouse ? no _

File: feelpp/feel/feelmesh/bvh.hpp [hunk 3]
```diff
@@ -306,273 +441,304 @@ namespace Feel
...
-        //     for(auto & primitive : M_primitiveInfo )
-        //     {
-        //         if (tree->checkIntersectionWithTriangle(rayon,primitive.M_primitiveNumber))
-        //             M_intersected_leaf.push_back(primitive.M_primitiveNumber);
-        //     }
-        // }
-
-        void traverse_stackless(BVHTree::BVHNode * tree, BVHRay const& rayon)
+
+    void traverse_stackless( BVH_InHouse::BVHNode * tree, ray_type const& rayon )
```
Comment (prudhomm) [inline feelpp/feel/feelmesh/bvh.hpp:605]: rename `traverseStackLess` ?

**Review by prudhomm — APPROVED** [review]

**Pull request merged**